Advertise onion messaging feature bit upon startup

orbitalturtle opened the pull request:
Closes #3 to advertise onion messaging feature bit on startup

One note that I wasn't totally sure if the PR you referenced meant that the bits 38 OR 39 need to be set, or both? Right now the PR sets 38 but can definitely change that if need be :)

carlaKC commented on src/main.rs:
```diff
@@ -87,3 +79,49 @@ fn parse_args() -> Result<LndCfg, ArgsError> {
+	    println!("Now setting the onion messaging feature bit...");
+	},
+        Err(status) => {
+            if !status.message().contains("invalid add action for bit 38, bit is already set") {
```
See comment above - to save us from the devilry of string matching, I think it woudl be easier to only call this if we haven't set the feature bit.

orbitalturtle commented on src/main.rs:
```diff
@@ -87,3 +79,49 @@ fn parse_args() -> Result<LndCfg, ArgsError> {
+async fn set_feature_bit(mut client: Client) {
+    let update = tonic_lnd::peersrpc::UpdateFeatureAction {
+        action: 0,
+        feature_bit: 38,
```
Ah got it, thanks for the link!

orbitalturtle commented on src/main.rs:
```diff
@@ -87,3 +79,49 @@ fn parse_args() -> Result<LndCfg, ArgsError> {
+        .await
+        .expect("failed to get info");
+
+    if !info.into_inner().features.contains_key(&38) {
```
Ah good question.

orbitalturtle commented on src/main.rs:
```diff
@@ -87,3 +79,49 @@ fn parse_args() -> Result<LndCfg, ArgsError> {
+/// does not currently persist the custom feature bits that are set via the RPC.
+async fn set_feature_bit(mut client: Client) {
+    let update = tonic_lnd::peersrpc::UpdateFeatureAction {
+        action: 0,
```
Good call looks like there is an updateaction enum

orbitalturtle replied:
Updated to use the enum. It seems like it won't take the enum directly and needs it converted to i32, which is maybe a bit confusing looking, but still seems more readable to me because at least it says "add" somewhere in there?

orbitalturtle commented:
Thanks for the review @carla ! 

I made those changes as well as a couple of others:
- I submitted another small PR to tonic_lnd to get the client to implement the "clone" trait, and am now using that version of the code in this PR. From what I've been reading, cloning the client is lightweight. Thought this might be useful for sending the client both to set_feature_bit and the signer and potentially elsewhere as well in the future as lndk grows.
- Returns a more helpful error message when a user forgets to make lnd with the peersrpc service set

Let me know what you think.

carlaKC commented on src/main.rs:
```diff
@@ -19,17 +23,24 @@ async fn main() {
     };
 
     let mut client = get_lnd_client(args).expect("failed to connect");
+    let mut client_clone = client.clone();
```
The `Lightning`/`Peers`/`Signer` clients already implement `clone()`, so you can use `client.lightning().clone` instead of requiring a further upstream change to `tonic_lnd`. 

Took me a minute to figure out how `tonic` is dealing with this all / why everything needs to be mutable, seems like it's unavoidable and best practise is to just [clone](https://github.com/carlaKC/lndk/pull/20#issuecomment-1466952243).

carlaKC commented on src/main.rs:
```diff
@@ -185,3 +196,53 @@ fn parse_args() -> Result<LndCfg, ArgsError> {
+                panic!("{}", SERVICE_UNIMPLEMENTED);
+            }
+
+            panic!("error updating node announcement {}", status);
```
nit: `status` can go into formatting string `{status}`

carlaKC commented on src/main.rs:
```diff
@@ -185,3 +196,53 @@ fn parse_args() -> Result<LndCfg, ArgsError> {
+/// Sets the onion messaging feature bit (described in this PR:
+/// https://github.com/lightning/bolts/pull/759/), to signal that we support
+/// onion messaging. This needs to be done every time we start up, because LND
+/// does not currently persist the custom feature bits that are set via the RPC.
+async fn set_feature_bit(mut client: Client) {
```
If we make this `client: & mut Client` then pass in `& mut client` above, I think that we get around some of the move issues / don't need to clone as much.

carlaKC replied:
I'd like to be able to test these functions without needing to pass around a full `Client` - I think that'll get annoying / verbose to mock. Could we define a trait that only has the functionality we need + pass in here as a trait impl so that it's easier to mock?

carlaKC commented on src/main.rs:
```diff
@@ -185,3 +196,53 @@ fn parse_args() -> Result<LndCfg, ArgsError> {
+        .peers()
+        .update_node_announcement(tonic_lnd::peersrpc::NodeAnnouncementUpdateRequest {
+            feature_updates,
+            color: String::from(""),
+            alias: String::from(""),
+            address_updates,
```
Can use `.Default::default()` do that we don't need to fill in empty values here.

orbitalturtle replied:
oh that's super nifty

carlaKC commented on src/main.rs:
```diff
@@ -185,3 +196,53 @@ fn parse_args() -> Result<LndCfg, ArgsError> {
+        panic!("onion messaging feature bit failed to be set")
+    }
+
+    println!("Successfully set onion messaging bit");
```
Adding `simple_logger::init_with_level(log::Level::Info).unwrap();` in #17, perhaps we can just copy the log setup in here and use `log` rather? That way this PR doesn't need to depend on the other, but we still get nice logging.

carlaKC commented on src/main.rs:
```diff
@@ -185,3 +196,53 @@ fn parse_args() -> Result<LndCfg, ArgsError> {
+    let onion_feature_bit = 39;
+    let update = tonic_lnd::peersrpc::UpdateFeatureAction {
+        action: i32::from(tonic_lnd::peersrpc::UpdateAction::Add),
+        feature_bit: onion_feature_bit,
```
Can we use the constant defined above here instead?

carlaKC commented on src/main.rs:
```diff
@@ -9,7 +9,11 @@ use std::cell::RefCell;
-use tonic_lnd::{Client, ConnectError};
+use tonic_lnd::{tonic::Code, Client, ConnectError};
+
+const ONION_FEATURE_BIT: u32 = 39;
```
nitpick: name `onion_messages_optional` (or something of the sort) to align with naming in the spec.

carlaKC commented on src/main.rs:
```diff
@@ -185,3 +196,53 @@ fn parse_args() -> Result<LndCfg, ArgsError> {
+        .expect("failed to get info");
+
+    if !info.into_inner().features.contains_key(&ONION_FEATURE_BIT) {
+        panic!("onion messaging feature bit failed to be set")
```
I think that we should return a `Result` here rather than panic, makes testing easier.

orbitalturtle replied:
So do you mean returning it as a result, then handling it similarly to this in main? https://github.com/carlaKC/lndk/blob/5af8c5d379028245c479dc817a63dd37db96dfea/src/main.rs#L16

carlaKC commented on src/main.rs:
```diff
@@ -9,7 +9,11 @@ use std::cell::RefCell;
+use tonic_lnd::{tonic::Code, Client, ConnectError};
+
+const ONION_FEATURE_BIT: u32 = 39;
+const SERVICE_UNIMPLEMENTED: &str = "Peer or sign service is unimplemented. Remember to
```
Can this specifically be a `PEER_SERVICE_UNIMPLEMENTED` because we'll know what rpc server we're missing when we make specific calls and they fail with `Unimplemented`. 

Also, wdyt about having an `ErrUnimplemented` enum that implements `Error` instead of a const? Not sure what's more idiomatic rust, but enums seem to be all the rage.

orbitalturtle replied:
Yes I like this idea, especially since there will be a few... signerrpc, peersrpc, dev and maybe others

carlaKC reviewed (commented):
Nice update! I think that we can get away without the followup change to `tonic_lnd`, which is nice. 

Comments mostly surrounding error handling + testability. Would like to add a trait + unit test to `set_feature_bit` mocking out the various error cases before merge.

orbitalturtle commented:
@carlaKC  When you get a chance for another look, I made those changes. :) I added traits passed into set_feature_bit to make mocking easier. And I added a series of tests for the various errors that can be thrown. 

Maybe there's a better way to organize the errors I grouped under SetOnionBitErrors. Lmk if you have any thoughts there... My main goal there was to group them so that any of those errors can be returned from set_feature_bit().

dunxen commented on src/main.rs:
```diff
@@ -238,3 +256,304 @@ fn parse_args() -> Result<LndCfg, ArgsError> {
+        // set_feature_bit should throw another error.
+        client_mock.expect_get_info().returning(|_| {
+            Ok(tonic_lnd::lnrpc::GetInfoResponse {
+                features: HashMap::from([(
+                    8,
+                    tonic_lnd::lnrpc::Feature {
+                        name: String::from("testing"),
+                        is_known: true,
+                        is_required: false,
+                    },
+                )]),
+                alias: Default::default(),
+                best_header_timestamp: Default::default(),
+                block_hash: Default::default(),
+                block_height: Default::default(),
+                chains: Default::default(),
+                color: Default::default(),
+                commit_hash: Default::default(),
+                identity_pubkey: Default::default(),
+                num_active_channels: Default::default(),
+                num_inactive_channels: Default::default(),
+                num_peers: Default::default(),
+                num_pending_channels: Default::default(),
+                require_htlc_interceptor: Default::default(),
+                synced_to_chain: Default::default(),
+                synced_to_graph: Default::default(),
+                testnet: Default::default(),
+                uris: Default::default(),
+                version: Default::default(),
```
nit: so a nice thing we can do here to reduce the boilerplate and also avoid any deprecation warnings is to use the [struct update syntax](https://doc.rust-lang.org/stable/book/ch05-01-defining-structs.html#creating-instances-from-other-instances-with-struct-update-syntax) with `Default::default()`
```suggestion
                features: HashMap::from([(
                    8,
                    tonic_lnd::lnrpc::Feature {
                        name: String::from("testing"),
                        is_known: true,
                        is_required: false,
                    },
                )]),
                ..Default::default()
```

orbitalturtle replied:
Ooohh this is awesome! Thanks for the tip

orbitalturtle replied:
Updated, looks so much cleaner B)

carlaKC commented on src/main.rs:
```diff
@@ -33,10 +36,25 @@ async fn main() {
         .get_info(tonic_lnd::lnrpc::GetInfoRequest {})
         .await
         .expect("failed to get info");
+    let info_clone = info.into_inner().clone();
+    let info_clone_2 = info_clone.clone();
```
I don't think we need either of these clones because we only borrow from `info`?

orbitalturtle replied:
So if we remove the clone, I can't access info.features on line 45, because info is moved on line 42 (because the use of into_inner() moves it, it seems). 

But to your point, I just cleaned this up and got rid of one of the clones since looks like that was unnecessary!

carlaKC commented on src/main.rs:
```diff
@@ -238,3 +256,125 @@ fn parse_args() -> Result<LndCfg, ArgsError> {
+        match self {
+            SetOnionBitError::UnimplementedPeersService => write!(
+                f,
+                "Peer service is unimplemented. Remember to
```
nit: let's go for lower case for `Display` + error strings, upper case for logs?

Going to add a `contributing.md` to spell this out, would be nice to at least attempt to be consistent in the beginning :')

orbitalturtle replied:
soudns good!

carlaKC commented on src/main.rs:
```diff
@@ -238,3 +256,125 @@ fn parse_args() -> Result<LndCfg, ArgsError> {
+    client: &mut impl NodeInfoRetriever,
+    peers_client: &mut impl UpdateNodeAnnouncement,
+) -> Result<(), SetOnionBitError> {
+    let update = tonic_lnd::peersrpc::UpdateFeatureAction {
+        action: i32::from(tonic_lnd::peersrpc::UpdateAction::Add),
+        feature_bit: ONION_MESSAGES_OPTIONAL as i32,
+    };
+    let feature_updates = vec![update];
```
Combine these two lines? I don't think it hurts readability too much.

carlaKC commented on src/main.rs:
```diff
@@ -238,3 +256,125 @@ fn parse_args() -> Result<LndCfg, ArgsError> {
+    peers_client
+        .update_node_announcement(tonic_lnd::peersrpc::NodeAnnouncementUpdateRequest {
+            feature_updates,
+            color: Default::default(),
+            alias: Default::default(),
+            address_updates: Default::default(),
```
Can be flattened into `..Default::default()` to auto-fill the remaining fields with defaults.

carlaKC commented on src/main.rs:
```diff
@@ -33,10 +36,25 @@ async fn main() {
+    let pubkey = PublicKey::from_str(&info_clone.identity_pubkey).unwrap();
     info!("Starting lndk for node: {pubkey}");
 
+    if !info_clone_2.features.contains_key(&ONION_MESSAGES_OPTIONAL) {
```
Add an `info` log indicating that we're going to attempt to set this feature bit? Would rather start verbose for easy debugging then cut back later :)

carlaKC commented on src/main.rs:
```diff
@@ -33,10 +36,25 @@ async fn main() {
+        let mut announcement_updater = UpdateNodeAnnClient {
+            peers_client: client.peers(),
+        };
+        match set_feature_bit(&mut node_info_retriever, &mut announcement_updater).await {
+            Ok(_) => {}
+            Err(err) => panic!("Error setting feaure bit: {err}"),
```
I've updated `main` to return `Result<()()>` in #20, can we duplicate that here? For the error case, we can `error!` log so that we can see the error string and then reutrn `Err()`
